core/backend/src/EntityReferences.ts

```diff
@@ -0,0 +1,131 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Schema
+ */
+
+import { ConcreteEntityTypes, ElementAspectProps, ElementProps, EntityReference, ModelProps } from "@itwin/core-common";
+import { Id64, Id64String } from "@itwin/core-bentley";
+import type { Entity } from "./Entity";
+import type { Model } from "./Model";
+import type { Element } from "./Element";
+import type { ElementAspect } from "./ElementAspect";
+import type { Relationship, RelationshipProps } from "./Relationship";
+import * as assert from "assert";
+
+/** Concrete classes that can be created, with the notable exception of CodeSpecs since it is does not derive from Entity
+ * other entity classes. In the future if there is a need
+ * to manage them this way, it can be added
+ * @alpha
+ */
+export type ConcreteEntity = Element | Model | ElementAspect | Relationship;
+
+/** Props for a [[ConcreteEntity]]
+ * @alpha
+ */
+export type ConcreteEntityProps = ElementProps | ModelProps | ElementAspectProps | RelationshipProps;
+
+/**
+ * Utilities for the [EntityReference]($common) type which is a kind of strings
+ * @alpha
+ */
+export namespace EntityReferences {
+  export function isModel(id: EntityReference) {
+    return id[0] === ConcreteEntityTypes.Model;
+  }
+  export function isElement(id: EntityReference) {
+    return id[0] === ConcreteEntityTypes.Element;
+  }
+  export function isElementAspect(id: EntityReference) {
+    return id[0] === ConcreteEntityTypes.ElementAspect;
+  }
+  export function isRelationship(id: EntityReference) {
+    return id[0] === ConcreteEntityTypes.Relationship;
+  }
+  export function toId64(id: EntityReference) {
+    return id.slice(1);
+  }
+
+  /** split a concrete entity id into its type and raw id */
+  export function split(id: EntityReference): [ConcreteEntityTypes, Id64String] {
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
+    return [id[0] as ConcreteEntityTypes, id.slice(1)];
+  }
+
+  /** used by the transformer to figure out where to check for the existence in a db of a concrete element id
+   * @internal
+   */
+  export function isValid(id: EntityReference): boolean {
+    return Id64.isValid(toId64(id));
+  }
+
+  /** create the invalid id for a concrete entity type
+   * @internal
+   */
+  export function makeInvalid(type: ConcreteEntityTypes): EntityReference {
+    return `${type}${Id64.invalid}`;
+  }
+
+  // necessary to prevent cyclic dependencies, the required modules will be in the require cache already so I don't store
+  /* eslint-disable @typescript-eslint/naming-convention,@typescript-eslint/no-var-requires */
```

Comment:
There's no less horrible way to solve this?

Reply:
I will experiment with moving this utility to a separate file. Only `ClassRegistry` uses it, but perhaps if I separate this to another file and import in a certain order I can remove the cycles.

Reply:
didn't work, the cycle is in `ClassRegistry<-Entity<-EntityReference<-Model<-Element<-Entity`
instead of the getters and dynamic requires in this file, I'm going to do just one dynamic require in  `Entity.collectReferenceIds`. Please comment on that solution here and resolve if you think it is better, I just pushed it.